eclipse-theia/theia
# Fix PseudoTerminal events

Signed-off-by: thegecko <rob.moran@arm.com>



#### What it does

Well, this is embarrassing. The [fix merged for opening terminals](https://github.com/eclipse-theia/theia/pull/10780) and [the subsequent revert](https://github.com/eclipse-theia/theia/pull/12098) have both failed to sufficiently fix the terminal opening issues.

It transpires that opening custom terminals via tasks and by manually creating them both have different code paths ending up with the pseudoterminal map tracking the terminals being keyed with different types of IDs.

PR https://github.com/eclipse-theia/theia/pull/12005 actually partially fixed this and led me to this PR (thanks @zvzuola), however the resize events and close events weren't fixed.

This PR allows process-based terminals (used by tasks and keyed by process ID) to be re-keyed by the tracked ID (from the widget ID), so all events correctly work for both types of terminal.

#### How to test


I've created a test extension over at https://github.com/thegecko/web-extension-tasks and the built VSIX can be found in this zip: [web-extension-tasks-1.0.0.vsix.zip](https://github.com/eclipse-theia/theia/files/10583241/web-extension-tasks-1.0.0.vsix.zip)

It exposes two commands:

- `Text: Run task` which [executes a task](https://github.com/thegecko/web-extension-tasks/blob/main/src/commands.ts#L12) using a custom terminal and writes out some text
- `Text: Show text` which [creates a custom terminal](https://github.com/thegecko/web-extension-tasks/blob/main/src/commands.ts#L38) and writes out some text

The extension also shows messages when the custom terminal is resized, closed or text is input.

To test, run both these commands and observe the text being displayed and events being raised when resized/closed.

Without this PR, only one command works and not all events are raised.

The issue has been confirmed and fixed on both the electron and browser examples.

#### Review checklist


#### Reminder for reviewers


- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

cc @arekzaluski
